velox/type/tests/TimestampConversionTest.cpp

```diff
@@ -104,6 +104,100 @@ TEST(DateTimeUtilTest, fromDateInvalid) {
       1970, 6, 31, "Date out of range: 1970-6-31"));
 }
 
+TEST(DateTimeUtilTest, fromWeekOfMonthDate) {
+  auto daysSinceEpochLenient =
+      [](int32_t year, int32_t month, int32_t weekOfMonth, int32_t dayOfWeek) {
+        auto result = util::daysSinceEpochFromWeekOfMonthDate(
+            year, month, weekOfMonth, dayOfWeek, true);
+        EXPECT_TRUE(!result.hasError());
+        return result.value();
+      };
+
+  EXPECT_EQ(4, daysSinceEpochLenient(1970, 1, 2, 1));
+  EXPECT_EQ(361, daysSinceEpochLenient(1971, 1, 1, 1));
+  EXPECT_EQ(396, daysSinceEpochLenient(1971, 2, 1, 1));
+
+  EXPECT_EQ(10952, daysSinceEpochLenient(2000, 1, 1, 1));
+  EXPECT_EQ(19905, daysSinceEpochLenient(2024, 7, 1, 1));
+
+  // Before unix epoch.
+  EXPECT_EQ(-3, daysSinceEpochLenient(1970, 1, 1, 1));
+  EXPECT_EQ(-2, daysSinceEpochLenient(1970, 1, 1, 2));
+  EXPECT_EQ(-31, daysSinceEpochLenient(1969, 12, 1, 1));
+  EXPECT_EQ(-367, daysSinceEpochLenient(1969, 1, 1, 1));
+  EXPECT_EQ(-724, daysSinceEpochLenient(1968, 1, 2, 1));
+  EXPECT_EQ(-719533, daysSinceEpochLenient(0, 1, 1, 1));
+
+  // Negative year - BC.
+  EXPECT_EQ(-719561, daysSinceEpochLenient(-1, 12, 1, 1));
+  EXPECT_EQ(-719897, daysSinceEpochLenient(-1, 1, 1, 1));
+
+  // Day in the previous month.
+  EXPECT_EQ(19783, daysSinceEpochLenient(2024, 2, 5, 5));
+  // Day in the next month.
+  EXPECT_EQ(19751, daysSinceEpochLenient(2024, 2, 1, 1));
+
+  // Out of range day of week.
+  EXPECT_EQ(338, daysSinceEpochLenient(1970, 12, 1, 0));
+  EXPECT_EQ(337, daysSinceEpochLenient(1970, 12, 1, -1));
+  EXPECT_EQ(337, daysSinceEpochLenient(1970, 12, 1, -8));
+
+  EXPECT_EQ(332, daysSinceEpochLenient(1970, 12, 1, 8));
+  EXPECT_EQ(333, daysSinceEpochLenient(1970, 12, 1, 9));
+  EXPECT_EQ(336, daysSinceEpochLenient(1970, 12, 1, 19));
+
+  // Out of range month.
+  EXPECT_EQ(-3, daysSinceEpochLenient(1970, 1, 1, 1));
+  EXPECT_EQ(207, daysSinceEpochLenient(1970, 8, 1, 1));
+  EXPECT_EQ(361, daysSinceEpochLenient(1970, 13, 1, 1));
+
+  EXPECT_EQ(-31, daysSinceEpochLenient(1970, 0, 1, 1));
+  EXPECT_EQ(-66, daysSinceEpochLenient(1970, -1, 1, 1));
+  EXPECT_EQ(-430, daysSinceEpochLenient(1970, -13, 1, 1));
+
+  // Out of range year.
+  auto result =
+      util::daysSinceEpochFromWeekOfMonthDate(292278995, 1, 1, 1, true);
+  EXPECT_EQ(result.error().message(), "Date out of range: 292278995-1-1");
```

Comment:
I thought we do not expect errors in lenient mode. This behavior doesn't seem to match the doc. Would you double check?

Reply:
You can see this conversation https://github.com/facebookincubator/velox/pull/10604#discussion_r1723223107. Maybe I should change the documentation.